File: Python/129.Sum Root to Leaf Numbers.py
#每个节点是一个个位数，从根节点到叶节点的一条路径组成一个大数，求所有这些数的和
#    1
#   / \
#  2   3     返回12+13=25
import copy
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sumNumbers(self, root):
        self.sum  = 0
        self.inorder(root,0)
        return self.sum

    def inorder(self,p,cursum):
        if p!=None:
            tmp = cursum*10 + p.val
            if p.left==None and p.right==None:
                self.sum += tmp
            else:
                self.inorder(p.left,tmp)
                self.inorder(p.right,tmp)


t1 = TreeNode(5)
t2 = TreeNode(4)
t3 = TreeNode(8)
t4 = TreeNode(1)
t5 = TreeNode(3)
t6 = TreeNode(4)
t7 = TreeNode(7)
t8 = TreeNode(2)
t9 = TreeNode(5)
t1.left = t2
t1.right = t3
t2.left = t4
t3.right = t6
t3.left = t5
t4.left = t7
t4.right = t8
t6.right = t9

s = Solution()
print(s.sumNumbers(t1))

File: C++/Tree/129.Sum Root to Leaf Numbers.cpp
/*
二叉树每个结点只包含数字0--9，每个从根到叶的路径代表一个整数。
返回所有路径代表整数的和。
*/
/*
思路：递归，上层结点乘10加到当前结点，一直往下传，直到根结点，加到sum里。
*/
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
private:
	int sum;
public:
    int sumNumbers(TreeNode* root) {
    	if(!root)	return 0;
    	sum = 0;
    	helpSumNumber(root,0);
        return sum;
    }
    void helpSumNumber(TreeNode* root,int presum)
    {
    	if(!root->left && !root->right)	sum += (presum+root->val);
    	if(root->left)	helpSumNumber(root->left,10*(root->val+presum));
    	if(root->right)	helpSumNumber(root->right,10*(root->val+presum));
    }
};
int main()
{
	Solution s;
	TreeNode root(1),t1(2),t2(3);
	root.right = &t1;
	t1.left = &t2;
	cout << s.sumNumbers(&root) << endl;
}

